---
title: "inspecting output data"
author: "Olivia Cattau"
date: "10/21/2021"
output: html_document
---
#load count data from kallisto output(s)
```{r}
countMatrix<-read.table(file="https://raw.githubusercontent.com/sr320/nb-2022/main/P_generosa/analyses/kallisto-0207.isoform.counts.matrix", header=TRUE, sep = '\t')
names(countMatrix)[1]<-"target_id"
head(countMatrix)
```

#load blast data, see blasting .sh file to see code (in GitHub)
```{r}
blast_data<-read.table(file="https://gannet.fish.washington.edu/gigas/data/p.generosa/Panopea-generosa-uniprot_blastx.tab")
names(blast_data)[1]<-"target_id" #renamed to match countMATRIX data
names(blast_data)[2]<-"identifiers" #renamed to search in swiss prot 
```

#Data maniupulation into long format
```{r}
library(tidyr)
simple_data<-countMatrix
long_data<- reshape(data=simple_data, idvar="target_id", varying=c("heart", "gonad", "ctenidia", "juv_sl", "juv_amb", "larvae"), v.name=c("tpm"), times=c("heart", "gonad", "ctenidia", "juv_sl", "juv_amb", "larvae"), new.row.names=1:8183754, direction="long")
names(long_data)[2]<-"tissue"
summary(long_data)
long_data$binary=ifelse(long_data$tpm > 0, 1, 0)
simple_data$heart=ifelse(simple_data$heart > 0, 1, 0)
simple_data$gonad=ifelse(simple_data$gonad > 0, 1, 0)
simple_data$ctenidia=ifelse(simple_data$ctenidia > 0, 1, 0)
simple_data$juv_amb=ifelse(simple_data$juv_amb > 0, 1, 0)
simple_data$juv_sl=ifelse(simple_data$juv_amb> 0, 1, 0)
simple_data$larvae=ifelse(simple_data$larvae > 0, 1, 0)
row.names(simple_data)<-simple_data$target_id
simple_data<-simple_data[,-1]
z<-as.matrix(simple_data)
```

#join identifiers to target_ids
```{r}
library(dplyr)
library(tidyr)
library(tidyverse)
#merging target_ids to count Matrix to get identifies for swissprot
generosa_long<-left_join(long_data, blast_data, by="target_id", all=TRUE)
generosa_long<-generosa_long %>% relocate(identifiers, .after = target_id)
generosa_long<-generosa_long[-c(6:15)]
generosa_long$identifiers<-sub("\\.\\d+$", "", generosa_long$identifiers) 
generosa_counts<-left_join(countMatrix, blast_data, by = "target_id", all=TRUE)
generosa_counts<-generosa_counts %>% relocate(identifiers, .after = target_id)
generosa_counts<-generosa_counts[-c(9:18)] # not every target ID has a protein identifier...could be an issue later on....
#generosa_counts[complete.cases(generosa_counts),] for now keep all NAs
generosa_counts$identifiers<-sub("\\.\\d+$", "", generosa_counts$identifiers) #removed decimal after identifiers so match Uniprot format
```

# Go to Uniprot website and grab all go terms
[Uniprot Website](https://www.uniprot.org/uniprot/?query=*&fil=reviewed%3Ayes#)
```{r}
#loaded into gannet and pulled down
uniprot_all_reviewed<-read.csv(file="https://gannet.fish.washington.edu/gigas/uniprot-reviewed_yes.tab", sep='\t', header=TRUE)
#uniprot_all_unreviewed<-""
```

# How to join tables by identifiers 
```{r}
names(uniprot_all_reviewed)[1]<-"identifiers" #renamed
pgenerosa_proteins<-left_join(generosa_counts, uniprot_all_reviewed, by ="identifiers")
pgenerosa_proteins_long<-left_join(generosa_long, uniprot_all_reviewed, by="identifiers")
write.csv(pgenerosa_proteins, file="/Users/oliviacattau/Documents/GitHub/code/characterize_larval_transciptome/pgenerosa_proteins_2_14_2022.tab")
```

# filter by tissue type 
```{r}
pgenerosa_heart<-pgenerosa_proteins %>% count(heart) #makes a table so you can see how many reads have 0 tpm, make a proportion
heart_zero<-max(pgenerosa_heart$n)
pgenerosa_gonad<-pgenerosa_proteins %>% count(gonad)
gonad_zero<-max(pgenerosa_gonad$n)
pgenerosa_ctenidia<-pgenerosa_proteins %>% count(ctenidia)
ctenidia_zero<-max(pgenerosa_ctenidia$n)
pgenerosa_larvae<-pgenerosa_proteins %>% count(larvae)
larvae_zero<-max(pgenerosa_larvae$n)
pgenerosa_juv_amb<-pgenerosa_proteins %>% count(juv_amb)
juv_amb_zero<-max(pgenerosa_juv_amb$n)
pgenerosa_juv_sl<-pgenerosa_proteins %>% count(juv_sl) #not sure what this is, maybe low OA exposure?
juv_sl_zero<-max(pgenerosa_juv_sl$n)
#proportion of zero reads to all reads 
all<-length(pgenerosa_proteins$target_id)
heart_expressed<-(1-(heart_zero/all))*100 #31.09%
gonad_expressed<-(1-(gonad_zero/all))*100 #18.45%
ctenidia_expressed<-(1-(ctenidia_zero/all))*100 #33.36%
larvae_expressed<-(1-(larvae_zero/all))*100 #36.29%
juv_amb_expressed<-(1-(juv_amb_zero/all))*100 #51.95%
juv_sl_expressed<-(1-(juv_sl_zero/all))*100 #52.20%
gene_expression<-matrix(c(heart_expressed, gonad_expressed, ctenidia_expressed, larvae_expressed, juv_amb_expressed, juv_sl_expressed), ncol=1, nrow=6, byrow=TRUE)
rownames(gene_expression)<-c("heart", "gonad", "ctenidia", "larvae", "juv amb", "juv sl")
colnames(gene_expression)<-c("% genes expressed by tissue type")
genes_expressed_by_tissue<-as.data.frame(gene_expression)
write.csv(genes_expressed_by_tissue, file="/Users/oliviacattau/Documents/GitHub/code/characterize_larval_transciptome/percent_genes_expressed_by_tissue.tab")
```

#calculate total efficiency of protein expression
```{r}
pgenerosa_proteins
total_identifiers<-length(pgenerosa_proteins$identifiers) #length of unsorted data
total_IDs<-length(pgenerosa_proteins$target_id) #length of all real Target_ID data
pgenerosa_complete_cases<-pgenerosa_proteins[complete.cases(pgenerosa_proteins), ]#removes all NAs in the matrix
all_complete_cases<-length(pgenerosa_complete_cases$target_id) #robust data set with GoTerms and Target-Ids
error_of_identification<-(all_complete_cases/total_IDs)*100 #5.2% success in identifying the protein expressed
```

# Further Data manipulation
```{r}
library(dbplyr)
###### Heart Data
pgenerosa_heart<-pgenerosa_proteins[-c(4:8)]
pgenerosa_heart<-pgenerosa_heart[order(pgenerosa_heart$heart, pgenerosa_heart$identifiers, na.last=TRUE, decreasing=TRUE),]
#get data
pgenerosa_heart<-subset(pgenerosa_heart, pgenerosa_heart$heart >0)
quantiles_heart<-as.data.frame(summarise(pgenerosa_heart, quantile=quantile(heart)))
print(quantiles_heart[4,1])
x<-22.13962
heart_q4<-subset(pgenerosa_heart, pgenerosa_heart$heart > x)
print(length(heart_q4$target_id))

##### Gonad Data
pgenerosa_gonad<-pgenerosa_proteins[-c(3,5:8)]
pgenerosa_gonad<-pgenerosa_gonad[order(pgenerosa_proteins$heart, na.last=TRUE, decreasing=TRUE),]
pgenerosa_gonad<-subset(pgenerosa_gonad, pgenerosa_gonad$gonad >0)
quantiles_gonad<-as.data.frame(summarise(pgenerosa_gonad, quantile=quantile(gonad)))
print(quantiles_gonad[4,1])
x<-69.8115
gonad_q4<-subset(pgenerosa_gonad, pgenerosa_gonad$gonad > x)
print(length(gonad_q4$target_id))

##### Ctenidia
pgenerosa_ctenidia<-pgenerosa_proteins[-c(3,4,6:8)]
pgenerosa_ctenidia<-pgenerosa_ctenidia[order(pgenerosa_proteins$ctenidia, na.last=TRUE, decreasing=TRUE),]
pgenerosa_ctenidia<-subset(pgenerosa_ctenidia, pgenerosa_ctenidia$ctenidia >0)
quantiles_ctenidia<-as.data.frame(summarise(pgenerosa_ctenidia, quantile=quantile(ctenidia)))
print(quantiles_ctenidia[4,1])
x<-2
ctenidia_q4<-subset(pgenerosa_ctenidia, pgenerosa_ctenidia$ctenidia > x)
print(length(ctenidia_q4$target_id))

##### Juvenile @ Ambient Conditions
juv_amb<-pgenerosa_proteins[-c(3:5,7,8)]
juv_amb<-juv_amb[order(pgenerosa_proteins$juv_amb, na.last=TRUE, decreasing=TRUE),]
juv_amb<-subset(juv_amb, juv_amb >0)
quantiles_juv_amb<-as.data.frame(summarise(juv_amb, quantile=quantile(juv_amb)))
print(quantiles_juv_amb[4,1])
x<-6
juv_amb_q4<-subset(juv_amb, juv_amb > x)
print(length(juv_amb_q4$target_id))

##### Juvenile @ Slightly OA conditions
juv_sl<-pgenerosa_proteins[-c(3:6,8)]
juv_sl<-juv_sl[order(pgenerosa_proteins$juv_sl, na.last=TRUE, decreasing=TRUE),]
juv_sl<-subset(juv_sl, juv_sl >0)
quantiles_juv_sl<-as.data.frame(summarise(juv_sl, quantile=quantile(juv_sl)))
print(quantiles_juv_sl[4,1])
x<-29
juv_sl_q4<-subset(juv_sl, juv_sl > x)
print(length(juv_sl_q4$target_id))

##### Larvae 
pgenerosa_larvae<-pgenerosa_proteins[-c(3:7)]
pgenerosa_larvae<-subset(pgenerosa_larvae, pgenerosa_larvae$larvae > 0)
quantiles_larvae<-summarise(pgenerosa_larvae, quantile=quantile(larvae))
print(quantiles_larvae[4,1])
x<-25.1232
larvae_q4<-subset(pgenerosa_larvae, pgenerosa_larvae$larvae > x)
print(length(larvae_q4$target_id))
```

#Find overlapping Data and make new data.frame
```{r}
library(dplyr)
all_data<-list(heart_q4, gonad_q4, ctenidia_q4, juv_amb_q4, juv_sl_q4, larvae_q4)
lapply(all_data, function(x) c(toString(dim(x)), toString(names(x))))
multi_merge<-merge(merge(merge(merge(merge(
  all_data[[1L]], 
  all_data[[2L]]), 
  all_data[[3L]]),
  all_data[[4L]]),
  all_data[[5L]]),
  all_data[[6L]])
all_overlapping<-length(multi_merge$target_id)
all_na<-as.data.frame(table(is.na(multi_merge$identifiers)))
overlapping_data_identifiers<-(all_na[1,2]) # all true values, not NAs
merged_ids<-as.data.frame(multi_merge$target_id)
success_of_identification<-overlapping_data_identifiers/all_overlapping
```

#use binary data to make Venn Diagram
```{r}
library(tidyverse)
long_data_zeros<-long_data[apply(long_data, 1, function(row) all(row !=0)),]
df<-long_data_zeros[-c(3,4)]
table_no_zeros<-as.data.frame(crossprod(table(df)))
write.csv(table_no_zeros, file="/Users/oliviacattau/Documents/GitHub/code/characterize_larval_transciptome/pgenerosa_table_no_zeros.tab")
simple_data_cleaned<-simple_data[apply(simple_data, 1, function(row) all(row !=0)),]
```


```{r}
#For Loop example, may not end up using 
attach(pgenerosa_cleaned)
vector<-numeric(length=length(pgenerosa_cleaned$est_counts))
for (i in seq_along(pgenerosa_cleaned$est_counts)) {
  vector[i] <- pgenerosa_cleaned$est_counts[i] >=10
}
head(vector)
```
